### ISSUE TITLE
Reading Timer Address Error

### ISSUE THREAD

**Issue by Zephiles** [FIX94/Nintendont#335]
This issue is with the game Paper Mario: The Thousand-Year Door.

The issue is that Nintendon't doesn't seem to read the value properly for the in-game timer. This value is stored at an 8 byte address, and in the English version of the game this address is located at 0x3DAC38. As an example, I hacked a save file to have an exact in-game time of 4:20 on the file select screen (this is hours/minutes). Under normal conditions, this value is stored as 0x000000931A36EE00. When Nintendon't reads this, however, the time comes up as 2:53. Since the time is only shown up to minutes, the value of 2:53 could be anywhere between 0x00000061E1325580 and 0x0000006272093100. So is this just Nintendon't reading the values wrong, or something else?

Also, this game uses a lot of 8 byte addresses like this. Since Nintendon't misreads this one, I would imagine it would misread the other addresses as well. This may lead to other issues during gameplay, and thus would probably need to be fixed in order for the game to function properly.

**Comment by GerbilSoft** [comment]
4:20 = 260 seconds
2:53 = 173 seconds
260 / 173 = ~1.501

Looks like a timer patch is being applied where it shouldn't be. Nintendont applies 1.5x timer patches to compensate for in-game timing due to the Wii's CPU being 1.5x faster than GameCube's, but that shouldn't be applied here.

**Comment by FIX94** [comment]
This is not nintendonts fault but your fault in your math, the wii processor runs 1.5x faster than the gamecube processor did, to calculate your value of 260 minutes you used this as a base:
https://github.com/FIX94/Nintendont/blob/master/kernel/PatchTimers.c#L63
but you should instead use this:
https://github.com/FIX94/Nintendont/blob/master/kernel/PatchTimers.c#L64
if you then add up all that you'll notice the displayed ingame time suddenly makes sense.

**Comment by Zephiles** [comment]
You're telling me this is my fault in math? I literally tested this file in both Dolphin and on a Wii without Nintendon't, and it works perfectly fine. And yet in Nintendon't it doesn't work properly. I don't see how this can possibly be my fault...

**Comment by FIX94** [comment]
a wii without nintendont=gamecube mode which clocks down the processor.

**Comment by Zephiles** [comment]
That doesn't mean that a some sort of patch couldn't be applied to properly read the value.

**Comment by TTYDSpeedrunner** [comment]
You're a fuckboi

**Comment by FIX94** [comment]
it DOES properly read the value though, its in processor ticks which will always be faster on wii, simple as that.

**Comment by Zephiles** [comment]
That doesn't change the fact that I can play on a Wii without Nintendon't and it reads the value differently than it does with Nintendon't.

**Comment by GerbilSoft** [comment]
This could be fixed with a patch for the display code; however, since this is merely a cosmetic issue that doesn't actually affect gameplay, it isn't a high priority.

**Comment by Zephiles** [comment]
I would say it could be a major issue, because someone could go from playing without Nintendon't to using it, in which the timers already set beforehand would be read differently. This can obviously cause some issues.

**Comment by GerbilSoft** [comment]
It's *not* read differently. The value is being formatted using the CPU clock speed, which is 1.5x faster in Wii mode. This only affects the timer display.

**Comment by Zephiles** [comment]
If that's the only thing being affected, then I don't see why a patch couldn't be applied to change that for this specific type of instance.

**Comment by GerbilSoft** [comment]
A patch could be applied. The difficulty is figuring out *where* to apply it. If you can reverse-engineer the executable and figure out where it's formatting the time display, we can add a patch for it.

**Comment by FIX94** [comment]
I hope you DO realize that doing so when you then actually play the game and NOT actually hack it the time would be displayed wrong since the processor tick rate increases it faster than it displays it for.

**Comment by Zephiles** [comment]
I don't have to hack a time to make this happen. Ive used unmodified save files with it plenty of times and this always happens.

**Comment by GerbilSoft** [comment]
Completely forgot about that. A new save that's only been on Nintendont would show the correct time, whereas the issue only shows up if you switch between Nintendont and MIOS/GameCube.

A more comprehensive patch would need to be written to adjust the internal timing code in that case.

**Comment by Zephiles** [comment]
I just got done doing a test with a friend of mine to see if this causes issues or not. 

A quick background is that I set up a file that is right before you would obtain a Yoshi from an egg, and the color is determined by the following: When you obtain the egg, the game sets the time that you get it. When the egg hatches, it subtracts that time from the in-game timer. It then uses some preset times to determine the color. Only 2 presets will be used for this test: 0-6 minutes for green, and 6-9 minutes for red. I set up a file so that exactly 4 minutes had passed after receiving the egg before saving the game. We then waited just over 2 minutes on both Nintendon't and without it via the disc channel to see what the color would be.

Upon performing the test on Nintendon't, the egg hatched after about 2 minutes 17 seconds, and the color was green. When tested on disc, the egg hatched after about 2 minutes 17 seconds, and the color was red. Obviously this suggests that this is in fact causing issues.
